Compact classnames for subcomponent implementations, fixes #421.

rjrjr opened the pull request:
Fix for https://github.com/google/dagger/issues/421.

Instead of including more and more levels of container names until we achieve
uniqueness, we try two strategies: use simple names everywhere, or else
include a uniquing prefix everywhere.

The prefix is built from the initials of a containing class for nested
subcomponent declarations, or the initials of containing packages for
top level. An integer may appended to the prefix if necessary.

ronshapiro commented:
I think once we get names like `F2_SubImpl` or `tdmltmt_SubImpl`, I think we've lost all value of having recognizeable names, and something like `{simple_name}{number}Impl` is probably just as good.

If we defined subcomponents as static classes, all nested one level inside the generated component, we can keep the more clear class names, but at the cost of needing to maintain our own pointers back to the parent components. There are other bugs that people have filed that want that (I think @pyricau had some).

rjrjr commented:
I tried simple names first, but they're just too long, we still had the same problem. WRT the readability of `F2_SubImpl`, test cases like that are not very realistic examples. In real code we see things like:

```
  @Override
  public SellerScope.PhoneComponent sellerScopePhone() {
    return new SS_PhoneComponentImpl();
  }

  @Override
  public SellerScope.TabletComponent sellerScopeTablet() {
    return new SS_TabletComponentImpl();
  }
```

and

```
  private final class SS_TabletComponentImpl implements SellerScope.TabletComponent {
```

In context I found it pretty readable -- I imagine that's why @gk5885 first suggested this approach. And such terms always appear next to something clickable for IDE navigation. 

I could take a stab at the static classes, but I'll need some handholding (and there's really something bothersome about reimplementing a language feature that way). It's your call, obviously, but I'd be happier landing this adequate solution.

rjrjr commented:
@ronshapiro Ping. Waiting to hear if you have been overwhelmed by the share elegance of my argument to merge this as-is, or if you can give me (very detailed) pointers on how to tackle the static class approach.

ronshapiro commented:
I think @gk5885 was more in favor of this then I was. The static class approach is likely to be a big undertaking, since there are a lot of parts that implicitly rely on the current structure. If once Greg chimes in we think it's worth considering, it probably makes more sense to discuss the static classes in a design doc format or GVC

gk5885 commented:
I like this strategy a lot -- there's really no ambiguity if the name doesn't collide and the source is likely to be as easy to navigate as it would be otherwise.  Ship it!

ronshapiro commented:
@gk5885 do you want to do a codereview here and then let me know and I'll pull it in?

gk5885 commented on compiler/src/main/java/dagger/internal/codegen/ComponentWriter.java:
```diff
@@ -103,36 +108,89 @@ private UniqueSubcomponentNamesGenerator(BindingGraph graph) {
-      for (int levels = 0; generatedSimpleNames.size() != components.size(); levels++) {
+
+      // Let's see if we can get away with using simpleName() everywhere.
+      for (ComponentDescriptor component : components) {
```
```
components
    .stream()
    .map(componentNamers::get)
    .map(Namer::simpleName)
    .allMatch(new HashSet<>()::add)
```
(or something like that) will tell you if you have a unique collection

rjrjr replied:
Sorry, my stream fu is weak. If this passes, won't it require me to iterate a second time to actually build the map?

gk5885 commented on compiler/src/main/java/dagger/internal/codegen/ComponentWriter.java:
```diff
@@ -103,36 +108,89 @@ private UniqueSubcomponentNamesGenerator(BindingGraph graph) {
+          String candidateName;
+          int disambiguate = 0;
+
+          do {
```
Can you use `dagger.internal.codegen.UniqueNameSet` for this?

rjrjr commented:
Belated thanks for the review!

I'm working on this now. Do you like updates squashed?

ronshapiro commented:
Don't worry about squashing. I'll take care of that when we merge
internally.

On Wed, Mar 22, 2017, 8:58 PM Ray Ryan <notifications@github.com> wrote:

> Belated thanks for the review!
>
> I'm working on this now. Do you like updates squashed?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/google/dagger/pull/606#issuecomment-288585775>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAwY3UDado8mXpg97YujH35ZFVKGhrEkks5rocO1gaJpZM4MRojO>
> .
>

ronshapiro commented:
@rjrjr @gk5885 is this ready to be submitted internally?

rjrjr commented:
@ronshapiro I was away, back now. Looks like I need to rebase, and I need @gk5885 guidance on how to do the stream thing at line 113.

rjrjr commented:
@ronshapiro This looks OK to merge to me if @gk5885 approves. I believe I've addressed all feedback except for the stream thing, I'm just not functional enough for that yet.

rjrjr commented:
Man, getting serious flashbacks doing these `bazel` builds.

ronshapiro commented on compiler/src/main/java/dagger/internal/codegen/ComponentWriter.java:
```diff
@@ -102,36 +108,77 @@ private UniqueSubcomponentNamesGenerator(BindingGraph graph) {
-            break;
-          }
+          Namer namer = componentNamers.get(component);
+          final String simpleName = namer.simpleName();
```
Can you remove `final` from this and the next line?

rjrjr replied:
Done, pushed.

rjrjr commented:
@gk5885 ping

ronshapiro commented:
This was actually submitted a few days ago in https://github.com/google/dagger/commit/ecba30a3fe3bf3e42c474ef5677f69f488569f53 but for some reason Github didn't auto-close it

ronshapiro commented:
Oh, I didn't list this PR also in the Fixes list, thats why.

rjrjr commented:
Wooo HOOOO! Thanks.

ronshapiro commented:
Thanks for contributing it! Hope you can now remove your long lived fork!

naturalwarren commented on compiler/src/main/java/dagger/internal/codegen/ComponentWriter.java:
```diff
@@ -102,36 +108,77 @@ private UniqueSubcomponentNamesGenerator(BindingGraph graph) {
+              component);
         }
       }
       return ImmutableBiMap.copyOf(generatedSimpleNames).inverse();
```
Won't this actually result in longer names when simple names are used now that we're not splitting the qualified name?

naturalwarren commented on compiler/src/main/java/dagger/internal/codegen/ComponentWriter.java:
```diff
@@ -102,36 +108,77 @@ private UniqueSubcomponentNamesGenerator(BindingGraph graph) {
         for (ComponentDescriptor component : components) {
-          List<String> pieces = componentQualifiedNamePieces.get(component);
-          String simpleName =
-              QUALIFIED_NAME_JOINER.join(
```
Won't this actually result in longer names when simple names are used now that we're not splitting the qualified name?

naturalwarren commented on compiler/src/main/java/dagger/internal/codegen/ComponentWriter.java:
```diff
@@ -102,36 +108,77 @@ private UniqueSubcomponentNamesGenerator(BindingGraph graph) {
+      ImmutableMap.Builder<ComponentDescriptor, Namer> builder = ImmutableMap.builder();
       for (ComponentDescriptor component : componentDescriptors) {
-        Name qualifiedName = component.componentDefinitionType().getQualifiedName();
-        builder.putAll(component, QUALIFIED_NAME_SPLITTER.split(qualifiedName));
```
Won't this actually result in longer names when simple names are used now that we're not splitting the qualified name?

ronshapiro commented:
Someone internally just ran into this, where each component had a different simple name and thus this renaming strategy didn't come into effect. We may need to implement another pass that checks whether or not the concatenated name of a nested subcomponent is too long, and therefore we should shorten names until we get under the threshold
